map2loop/utils.py

```diff
@@ -0,0 +1,302 @@
+    dipdir = numpy.degrees(numpy.arctan2(normal_vector[:, 0], normal_vector[:, 1])) % 360
+
+    # Calculate the dip angle in degrees, ranging from 0 to 90
+    dip = 90 - numpy.degrees(numpy.arcsin(normal_vector[:, 2]))
+
+    # If the dip angle is greater than 90 degrees, adjust the dip and dip direction
+    mask = dip > 90
+    dip[mask] = 180 - dip[mask]
+    dipdir[mask] = (dipdir[mask] + 180) % 360
+
+    # Ensure the dip direction is within the range of 0 to 360 degrees
+    dipdir = dipdir % 360
+    dip_dipdir = numpy.array([dip, dipdir]).T
+
+    return dip_dipdir
+
+
+@beartype.beartype
+def create_points(xy: Union[list, tuple, numpy.ndarray]):
+    """
+    Creates a list of shapely Point objects from a list, tuple, or numpy array of coordinates.
+
+    Args:
+        xy (Union[list, tuple, numpy.ndarray]): A list, tuple, or numpy array of coordinates,
+        where each coordinate contains two elements representing the x and y coordinates of a point.
+
+    Returns:
+        shapely.points: A list of Point objects created from the input list of coordinates.
+    """
+    points = shapely.points(xy)
+    return points
+
+
+@beartype.beartype
+def find_segment_strike_from_pt(
+    line: shapely.LineString, point: shapely.Point, measurement: pandas.Series
+) -> float:
+    """
+    Finds the strike of a line segment (contact) closest to a given point (structural measurement).
+
+    Parameters:
+    line (shapely.geometry.LineString): The line segment (from sampled_contacts)
+    point (shapely.geometry.Point): The point representing the structural measurement
+    measurement (Geoseries): sampled_contacts.iloc[i, :], where i is a given row number.
+
+    Returns:
+    float: The strike of the line segment closer to the point
+    """
+
+    lines = []
+    for c1, c2 in zip(line.coords, line.coords[1:]):
+        lines.append(shapely.LineString([c1, c2]))
+    distances = [segment.distance(point) for segment in lines]
+    nearest_line = lines[distances.index(min(distances))]
+
+    if 0 <= measurement['DIPDIR'] <= 180:
+        # 1 is the upper point
+        # find the index point in seg a that has highest y value
+        idx1 = numpy.argmin(nearest_line.coords.xy[1])
+        x1 = nearest_line.coords.xy[0][idx1]
+        y1 = nearest_line.coords.xy[1][idx1]
+        idx2 = numpy.argmax(nearest_line.coords.xy[1])
+        x2 = nearest_line.coords.xy[0][idx2]
+        y2 = nearest_line.coords.xy[1][idx2]
+
+    if 180 < measurement['DIPDIR'] <= 360:
+        # 1 is the lower point
+        idx1 = numpy.argmax(nearest_line.coords.xy[1])
+        x1 = nearest_line.coords.xy[0][idx1]
+        y1 = nearest_line.coords.xy[1][idx1]
+        idx2 = numpy.argmin(nearest_line.coords.xy[1])
+        x2 = nearest_line.coords.xy[0][idx2]
+        y2 = nearest_line.coords.xy[1][idx2]
+
+    strike = numpy.degrees(math.atan2((x2 - x1), (y2 - y1))) % 360
+    return strike
+
+
+@beartype.beartype
+def calculate_endpoints(
```

Comment:
Because this is the bounding box of the polygon in which the structural datapoint is in. 
The bounds are added into the geology geodataframe as columns, so it's just easier to call the bbox this way.